planner/util/path.go

```diff
@@ -181,3 +181,105 @@ func (path *AccessPath) OnlyPointRange(sctx sessionctx.Context) bool {
 	}
 	return noIntervalRange && !haveNullVal
 }
+
+// Col2Len maps expression.Column.UniqueID to column length
+type Col2Len map[int64]int
+
+// ExtractCol2Len collects index/table columns with lengths from expressions. If idxCols and idxColLens are not nil, it collects index columns with lengths(maybe prefix lengths).
+// Otherwise it collects table columns with full lengths.
+func ExtractCol2Len(exprs []expression.Expression, idxCols []*expression.Column, idxColLens []int) Col2Len {
+	col2len := make(Col2Len, len(idxCols))
+	for _, expr := range exprs {
+		extractCol2LenFromExpr(expr, idxCols, idxColLens, col2len)
+	}
+	return col2len
+}
+
+func extractCol2LenFromExpr(expr expression.Expression, idxCols []*expression.Column, idxColLens []int, col2Len Col2Len) {
+	switch v := expr.(type) {
+	case *expression.Column:
+		if idxCols == nil {
+			col2Len[v.UniqueID] = types.UnspecifiedLength
+		} else {
+			for i, col := range idxCols {
+				if v.UniqueID == col.UniqueID {
+					col2Len[v.UniqueID] = idxColLens[i]
+					break
+				}
+			}
+		}
+	case *expression.ScalarFunction:
+		for _, arg := range v.GetArgs() {
+			extractCol2LenFromExpr(arg, idxCols, idxColLens, col2Len)
+		}
+	}
+}
+
+func compareLength(l, r int) int {
+	if l == types.UnspecifiedLength && r == types.UnspecifiedLength {
+		return 0
+	}
+	if l == types.UnspecifiedLength {
+		return 1
+	}
+	if r == types.UnspecifiedLength {
+		return -1
+	}
+	if l > r {
+		return 1
+	}
+	if l == r {
+		return 0
+	}
+	return -1
+}
+
+// c1 dominates c2 when each column of c2 exists in c1 and c2's column length is no longer than c1's column length.
```

Comment:
Maybe we should follow the pattern
```suggestion
// dominates checks/tells ...
```